### PR TITLE
RHOAIENG-53866: Add kustomize variable support for Workbenches component

### PR DESCRIPTION
**Pull request by harshad16**
## Description

This PR adds support for dynamic kustomize variables in the Workbenches component, similar to the existing implementation in MLflowOperator and Dashboard components.
#### Changes

- `workbenches_support.go`
    - Added ComputeKustomizeVariable function that computes:
        - gateway-url: The gateway domain URL (e.g., https://odh-gateway.apps.example.com/)
        - section-title: Platform-specific title (OpenShift Open Data Hub / OpenShift Self Managed Services / OpenShift Managed Services)
        - mlflow-enabled: "true" or "false" based on whether MLflowOperator is managed in the DataScienceCluster
    - Added isMLflowEnabled helper function to check MLflowOperator management state from DSC
    - Added sectionTitle map for platform-specific titles
- `workbenches_controller_actions.go`
    - Added setKustomizedParams action that applies the computed variables to the notebook controller's params.env
- `workbenches_controller.go`
    - Registered setKustomizedParams action in the reconciler chain
- `workbenches_support_test.go (new file)`
    - Added comprehensive tests for ComputeKustomizeVariable covering:
        - Different platforms (OpenDataHub, SelfManagedRhoai, ManagedRhoai)
        - MLflow enabled/disabled scenarios
        - Error handling
- `tests/e2e/workbenches_test.go`
    - Added new test case ValidateMLflowIntegration that:
    - Verifies MLflowOperator management state is accessible
    - Ensures Workbenches component remains ready with MLflowOperator state
    - Validates the notebook controller deployment is available

### Summary

  - Added ComputeKustomizeVariable function to compute dynamic kustomize variables for the notebook controller manifests
  - Added setKustomizedParams reconciler action to apply the computed variables to params.env
  - Variables include gateway URL, section title (platform-specific), and MLflow enabled status

This PR is generated with help of `Cursor - Claude-4.5-opus model`

### Test plan

[x] Unit tests pass (go test ./internal/controller/components/workbenches/... -v -run TestComputeKustomizeVariable)
[x] Verify params.env is correctly updated during reconciliation
[x] Verify the kustomize variables are properly substituted in the manifests

Related-to: https://redhat.atlassian.net/browse/RHOAIENG-53866

## How Has This Been Tested?

Tested in openshift cluster with following commands:

```
make image-build image-push IMG=quay.io/harshad16/opendatahub-operator:rhoai-3.4

make bundle bundle-build bundle-push \
  IMAGE_TAG_BASE=quay.io/harshad16/opendatahub-operator \
  IMG_TAG=rhoai-3.4 \
  BUNDLE_IMG=quay.io/harshad16/opendatahub-operator-bundle:rhoai-3.4 \
  ODH_PLATFORM_TYPE=rhoai

./bin/operator-sdk cleanup rhods-operator -n redhat-ods-operator
./bin/operator-sdk run bundle quay.io/harshad16/opendatahub-operator-bundle:rhoai-3.4 \
  --namespace redhat-ods-operator \
  --decompression-image quay.io/project-codeflare/busybox:1.36
```

## Screenshot or short clip

When mlflow-operator: managedState: Removed

Screenshot from 2026-03-20 04-55-55
Screenshot from 2026-03-20 02-13-17

When mlflow-operator: managedState: Managed

Screenshot from 2026-03-20 04-41-19
Screenshot from 2026-03-20 04-40-59

## Merge criteria

- [x] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [x] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [x] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has manually tested the changes and verified that the changes work
- [x] The developer has run the integration test pipeline and verified that it passed successfully
- [x] New RELATED_IMAGE mappings are already listed in ODH-Build-Config and RHOAI-Build-Config, and links are included in PR description

### E2E test suite update requirement

When bringing new changes to the operator code, such changes are by default required to be accompanied by extending and/or updating the E2E test suite accordingly.

To opt-out of this requirement:
1. **Please inspect the [opt-out guidelines](https://github.com/opendatahub-io/opendatahub-operator/blob/main/docs/e2e-update-requirement-guidelines.md)**, to determine if the nature of the PR changes allows for skipping this requirement
2. If opt-out is applicable, provide justification in the dedicated `E2E update requirement opt-out justification` section below
3. Check the checkbox below:
- [ ] Skip requirement to update E2E test suite for this PR
4. Submit/save these changes to the PR description. This will automatically trigger the check.

## Summary by CodeRabbit

* **New Features**
  * MLflow integration for Workbenches with automatic enablement detection from cluster state
  * Workbenches now apply computed kustomize parameters, set gateway URL, and show platform-specific section titles

* **Tests**
  * Unit tests validating kustomize variable computation, validation, and edge cases
  * End-to-end test validating MLflow enablement toggling and Workbenches readiness

### CODE REVIEW

File: internal/controller/components/workbenches/workbenches_controller.go
Comment (carlkyrillos) [inline internal/controller/components/workbenches/workbenches_controller.go:64]:
```suggestion
		Watches(&corev1.Namespace{}).
		// Watch MLflowOperator so that changes to its management state (enabled/disabled)
		// trigger a Workbenches reconciliation to update the mlflow-enabled param.
		WatchesGVK(gvk.MLflowOperator,
			reconciler.WithEventHandler(handlers.ToNamed(componentApi.WorkbenchesInstanceName)),
			reconciler.WithPredicates(predicates.DefaultPredicate),
		).
```
You should add a Watch on the `MLflowOperator` CR so the workbenches controller will reconcile whenever MLflow is enabled/disabled.

Comment (harshad16) [reply internal/controller/components/workbenches/workbenches_controller.go:64]: Thank you so much for sharing this, it indeed would help notebook-controller, will incorporate it.

File: internal/controller/components/workbenches/workbenches_support.go [hunk 1]
```diff
@@ -80,3 +94,35 @@ func notebookImagesManifestInfo(sourcePath string) odhtypes.ManifestInfo {
...
+		return nil, fmt.Errorf("error getting gateway domain: %w", err)
+	}
+
+	mlflowEnabled, err := isMLflowEnabled(ctx, cli)
+	if err != nil {
+		return nil, fmt.Errorf("error checking MLflow status: %w", err)
+	}
+
+	return map[string]string{
+		"gateway-url":    fmt.Sprintf("https://%s/", consoleLinkDomain),
```
Comment (ugiordan) [inline internal/controller/components/workbenches/workbenches_support.go]: The `consoleLinkDomain` value is written directly into `params.env` without sanitization. Since `params.env` is a line-based key=value format, a domain containing a newline character (e.g., `evil.com\nmalicious-key=malicious-value`) could inject arbitrary key-value pairs into the file.

While the domain originates from cluster-level CRs (requiring admin privileges to set), defense-in-depth would suggest validating the domain before use, at a minimum rejecting strings containing `\n`, `\r`, or `=`.

Something like:
```
if strings.ContainsAny(consoleLinkDomain, "\n\r=") {
   return nil, fmt.Errorf("invalid gateway domain %q: contains illegal characters", consoleLinkDomain)
}
```

File: internal/controller/components/workbenches/workbenches_support.go [hunk 2]
```diff
@@ -80,3 +94,35 @@ func notebookImagesManifestInfo(sourcePath string) odhtypes.ManifestInfo {
...
+		"section-title":  sectionTitle[platform],
```
Comment (ugiordan) [inline internal/controller/components/workbenches/workbenches_support.go]: If `platform` doesn't match any key in the `sectionTitle` map, this silently returns an empty string. Consider adding a fallback or returning an error for unknown platforms to avoid writing `section-title=` (empty value) into `params.env`.

File: internal/controller/components/workbenches/workbenches_support.go [hunk 3]
```diff
@@ -80,3 +94,35 @@ func notebookImagesManifestInfo(sourcePath string) odhtypes.ManifestInfo {
...
+	dscList := &dscv2.DataScienceClusterList{}
+	if err := cli.List(ctx, dscList); err != nil {
+		return false, fmt.Errorf("failed to list DataScienceClusters: %w", err)
+	}
+
+	if len(dscList.Items) == 0 {
+		return false, nil
+	}
+
+	dsc := &dscList.Items[0]
```
Comment (ugiordan) [inline internal/controller/components/workbenches/workbenches_support.go]: This takes the first `DSC` from an unordered list. I know the convention is a singleton `DSC`, but if this is a pattern used elsewhere in the codebase, it's fine. I'm just calling it out for awareness.

File: internal/controller/components/workbenches/workbenches_support.go [hunk 4]
```diff
@@ -80,3 +94,35 @@ func notebookImagesManifestInfo(sourcePath string) odhtypes.ManifestInfo {
...
+
+	return map[string]string{
+		"gateway-url":    fmt.Sprintf("https://%s/", consoleLinkDomain),
+		"section-title":  sectionTitle[platform],
+		"mlflow-enabled": strconv.FormatBool(mlflowEnabled),
+	}, nil
+}
+
+func isMLflowEnabled(ctx context.Context, cli client.Client) (bool, error) {
+	dscList := &dscv2.DataScienceClusterList{}
```
Comment (lburgazzoli) [inline internal/controller/components/workbenches/workbenches_support.go]: there should be a `cluster.GetDSC(ctx, cli)` that wouls simplify this

File: internal/controller/components/workbenches/workbenches_support.go (hunk 1 above)
Comment (harshad16) [reply internal/controller/components/workbenches/workbenches_support.go]: very viable thought, i have included this in the code base

File: internal/controller/components/workbenches/workbenches_support.go (hunk 2 above)
Comment (harshad16) [reply internal/controller/components/workbenches/workbenches_support.go]: thanks, adjusted the changes as per direction.

File: internal/controller/components/workbenches/workbenches_support.go (hunk 3 above)
Comment (harshad16) [reply internal/controller/components/workbenches/workbenches_support.go]: Adjust the changes, based on suggestion given here:
https://github.com/opendatahub-io/opendatahub-operator/pull/3293#discussion_r2960979596
new changes are in the code-base

File: internal/controller/components/workbenches/workbenches_support.go (hunk 4 above)
Comment (harshad16) [reply internal/controller/components/workbenches/workbenches_support.go]: made the new changes, to use this over the changes.
thank you

**Comment by harshad16** [comment]
Tested the changes on the cluster
included the details in the description.

Ready for review

File: internal/controller/components/workbenches/workbenches_support.go [hunk 5]
```diff
@@ -80,3 +95,42 @@ func notebookImagesManifestInfo(sourcePath string) odhtypes.ManifestInfo {
...
+
+func ComputeKustomizeVariable(ctx context.Context, cli client.Client, platform common.Platform) (map[string]string, error) {
+	consoleLinkDomain, err := gateway.GetGatewayDomain(ctx, cli)
+	if err != nil {
+		return nil, fmt.Errorf("error getting gateway domain: %w", err)
+	}
+
+	if strings.ContainsAny(consoleLinkDomain, "\n\r=") {
+		return nil, fmt.Errorf("invalid gateway domain %q: contains illegal characters", consoleLinkDomain)
+	}
```
[the remark this answers is not in the document]
Comment (harshad16) [reply internal/controller/components/workbenches/workbenches_support.go]: Thank you for the changes suggested, adjusted in: https://github.com/opendatahub-io/opendatahub-operator/commit/24b644c7c6e2ed8b2875ba62d79d7b6c43de5239

File: tests/e2e/workbenches_test.go [hunk 6]
```diff
@@ -71,3 +73,35 @@ func (tc *WorkbenchesTestCtx) ValidateWorkbenchesNamespaceConfiguration(t *testi
...
+	)
+
+	// Verify the notebook controller deployment exists and is available
+	tc.EnsureResourceExists(
+		WithMinimalObject(gvk.Deployment, types.NamespacedName{
+			Name:      "notebook-controller-deployment",
+			Namespace: tc.AppsNamespace,
+		}),
+		WithCondition(jq.Match(`.status.conditions[] | select(.type == "Available") | .status == "True"`)),
+	)
```
[the remark this answers is not in the document]
Comment (harshad16) [reply tests/e2e/workbenches_test.go:106]: Addressed in: https://github.com/opendatahub-io/opendatahub-operator/commit/611f2d4b7c52d1c333a32476c4e2f467419705e3
Thanks for the suggestioned 
Tested teh e2e: `make e2e-test-single TEST="TestOdhOperator/Component_Tests/workbenches/Validate MLflow integration"`
